### PR TITLE
wrap wagtail publish/unpublish/delete actions with transactions

### PR DESCRIPTION
**Pull request by mark-carrington**
# notes #

This change wraps the publish_action method in create_page.py and edit_page.py with the @transaction.atomic decorator, to ensure we roll back any DB changes if anything goes wrong. The publish_action method covers the signal handling before and after the publish, which is where our cache invalidation logic is run. With this we can ensure we roll back any DB changes associated with the publish if any of our cache invalidation logic doesn't complete successfully.

I have tired to place @transaction.atomic at a high level, as I want to associated entities (which could be removed as part of a publish) to part of the same transaction as the publish itself. Due to this, I've covered the following:
* create page (publish)
* edit page (publish)
* publish_scheduled_pages management command (publish)
* delete page (delete)
* unpublish page (unpublish)

### CODE REVIEW

**Review by Abdul-Dridi — APPROVED** [review]

File: wagtail/admin/tests/pages/test_create_page.py [hunk 1]
```diff
@@ -57,6 +58,44 @@ def test_add_subpage(self):
         # List of available page types should not contain pages whose parent_page_types forbid it
         self.assertNotContains(response, "Business child")
 
+    @mock.patch("wagtail.admin.views.pages.create.transaction.atomic", wraps=transaction.atomic)
+    def test_create_page_publish_transaction_rolled_back_on_error(self, mock_atomic):
```
Comment (pykita) [inline wagtail/admin/tests/pages/test_create_page.py:62]: nice test!

**Review by pykita — APPROVED** [review]
Nice one! Very elegant solution!

File: wagtail/admin/views/pages/create.py [hunk 2]
```diff
@@ -189,6 +190,7 @@ def save_action(self):
         # remain on edit page for further edits
         return self.redirect_and_remain()
 
+    @transaction.atomic
```
Comment (matt-dalton) [inline wagtail/admin/views/pages/create.py:193]: OOI have you seen any commentary online why they don't make this atomic? Just wondering if there could be a downside

File: wagtail/models/__init__.py [hunk 3]
```diff
@@ -2824,6 +2824,7 @@ def delete(self):
 
         return super().delete()
 
+    @transaction.atomic
```
Comment (matt-dalton) [inline wagtail/models/__init__.py:2827]: i guess if our caching is (to an extent) now reliant on forked wagtail functionality, we should definitely document this when we wrap up

**Review by matt-dalton — APPROVED** [review]
tests look good!

File: wagtail/admin/views/pages/create.py (hunk 2 above)
Comment (mark-carrington) [reply wagtail/admin/views/pages/create.py:193]: There's a performance cost to making a wider range of DB operations part of the same transaction I believe. Some parts of wagtail do use transactions, but the scope seems to be limited. I have read that nested transactions are fine, so it shouldn't cause any issues: https://docs.djangoproject.com/en/5.0/topics/db/transactions/

File: wagtail/models/__init__.py (hunk 3 above)
Comment (mark-carrington) [reply wagtail/models/__init__.py:2827]: yeah, will make sure this is documented

**Pull request merged**